File: question/leetcode/leetcode0019-Remove-Nth-Node-From-End-of-List/leetcode0019.js
function removeNthFromEnd(head, n) {
  let dummy = new ListNode(-1);
  dummy.next = head;
  let target = findNthFromEnd(dummy, n + 1);
  target.next = target.next.next;
  return dummy.next;
}

function findNthFromEnd(head, n) {
  let fast = head;
  for (let i = 0; i < n; i++) {
    fast = fast.next;
  }
  let slow = head;
  while (fast !== null) {
    fast = fast.next;
    slow = slow.next;
  }
  return slow;
}

File: question/leetcode/leetcode0019-Remove-Nth-Node-From-End-of-List/leetcode0019.py
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def removeNthFromEnd(head, n):
    dummy = ListNode(0)
    dummy.next = head
    target = findNthFromEnd(dummy, n+1)
    target.next = target.next.next
    return dummy.next


def findNthFromEnd(head, n):
    if not head:
        return None
    fast = slow = head
    for i in range(n):
        if not fast:
            return None
        fast = fast.next
    while fast:
        fast = fast.next
        slow = slow.next
    return slow